### PR TITLE
Custom Block Icons

### LINKED ISSUE #8719: Iterations on block library
I took a little trip into imagining some minimal iterations to the block library that hopefully solve a few issues.

The first issue is that right now we have a limited selection of icons to choose from for blocks, and it's causing the block library to lose clarity very rapidly once the user starts installing plugins that offer new blocks. The solution here employs custom icons (pulling from Material icons) into the mix along with Dashicons, we can work out which are used where. This has a huge benefit for those that are making blocks, you now can draw on a far wider range of options for your block on top of supplying your own SVG.

In addition to adding icons I also looked at adding more space to the experience. This brings in easier understanding and over all just makes it all feel a bit comfier to interact with.

20px

The final fix in this suggestion is to bring in a different style for child blocks. This stacking visual I think works in the context of these changes.

I am interested in what people think.

### PR DESCRIPTION
**Pull request by mtias**
_In Progress_

This update seeks to add custom icons to the block library in order to reduce the expectation of dashicon usage on blocks, which has resulted in a loss of clarity as soon as multiple blocks share the same icon (quote and pullquote, for example). Dashicons were also designed for UI affordance while blocks have distinct qualities that we should reflect.

It also hopes to encourage people to supply a custom SVG for their own blocks.

Closes #8719.

image

The new icons have a 24px base grid (pulling from Material design). All core blocks have been replaced except for a few where we had custom icons already which need updating to the new sizing (shortcode and custom HTML). This also reworks all the icons used for embeds and favors service specific icons when available for clarity.

## Missing Icons

- [x] Shortcode.
- [x] HTML.
- [x] Code.
- [x] Headings.
- [x] Inline Image.
- [x] Cover Image.

## Remaining Tasks

- [x] Address slash inserter.
- [x] Implement stacked icon deck.
- [x] Work on hover styles.
- [ ] Define spread of background color.

### CODE REVIEW

**Comment by chrisvanpatten** [comment]
I love that this uses raw SVGs instead of some kind of special icon name thing. Encouraging raw SVGs is great.

Hopefully this also opens the door to improving those icons without worrying about matching an icon suite necessarily — that quote icon, for instance, feels like it could use a little love. Because it's raw SVG I imagine it'll be easier to make small case-by-case tweaks rather than trying to get Google to approve a design change upstream.

Super cool!

**Comment by youknowriad** [comment]
screen shot 2018-08-13 at 13 03 37

I think it's not easy to distinguish whether the title applies to the icon above it or under it. I think we can fix this by reducing the padding of the buttons and increasing the margin bottom of the items. (Something like the screenshot above)

Also, I don't like the fact that the scrollbars show up surrounded by a padding. I think it's better to avoid the global padding and move it to each child div instead, that way the scrollbars show up at the edge.

File: packages/block-library/src/image/index.js [hunk 1]
```diff
@@ -89,12 +89,30 @@ const schema = {
...
+			<g id="ui_x5F_spec_x5F_header_copy_4">
+			</g>
+			<g>
+				<path id="XMLID_53_" d="M19,5v14H5V5H19 M19,3H5C3.9,3,3,3.9,3,5v14c0,1.1,0.9,2,2,2h14c1.1,0,2-0.9,2-2V5C21,3.9,20.1,3,19,3    L19,3z"/>
+				<path id="XMLID_51_" d="M14.14,11.86l-3,3.87L9,13.14L6,17h12L14.14,11.86L14.14,11.86z"/>
+			</g>
+		</g>
+	</svg>
+);
+
```
Comment (paulwilde) [inline packages/block-library/src/image/index.js]: This SVG seems to have a bunch of unnecessary IDs and <g> tags.

**Comment by ccprog** [comment]
I just came across this PR and noted that some of the SVGs use the following pattern:

    <svg width="24" height="24"><path fill="none" d="M0 0h24v24H0V0z"/>...

Note that there is no `viewBox` attribute, but instead an empty "backround" path element covering the whole viewport. If an icon defined like this is embeded as a ``, its size can be styled. But if it is used inlined in HTML, changing the size with CSS does not work. You would be well-advised to use

    <svg viewBox="0 0 24 24">...

for all icons to take advantage of SVGs intrinsic responsiveness, both to enable theming and for the sake of maintainablility. (All you'll ever have to do is adjust sizes in CSS.)

As an aside none of the other attributes sprinkled through there are really needed:

- `width` and `height` should be and are provided by CSS,
- `version` is never evaluated by browsers,
- `enable-background` is deprecated,
- `xmlSpace` is a valid attribute - in JSX -, but is superfluous for these icons, as they have no text content
- `xmlns="http://www.w3.org/2000/svg"` is a matter of convention. It is not strictly needed for `<svg>` elements inlined in HTML, but if you ever copy-and-paste them to a file that is evaluated as XML (like a standalone SVG file), it is compulsory.

**Comment by chrisvanpatten** [comment]
It might be helpful to run the icons through `svgo`, which can help normalise some of the issues that @ccprog mentioned. I usually use this nice GUI option: https://jakearchibald.github.io/svgomg/

**Comment by ccprog** [comment]
@chrisvanpatten SVGO is a heavily opinonated tool. In this case, it would _remove_ the viewBox as a default. Despite what the GUI seems to claim, it cannot _add_ a missing viewBox. Option "Prefer viewBox to width/height" only means it will dellete width/height if a viewBox is present. If it is not, a conscious designer decision is needed.

**Comment by jasmussen** [comment]
Pushed a little polish to the block chip, and the inline image icon:

screen shot 2018-08-14 at 10 07 49

**Comment by youknowriad** [comment]
What impact this PR should have on the block switcher's icon and the block style variations panel? (which mimic the inserter design)

**Comment by jasmussen** [comment]
Pushed a few fixes:

screen shot 2018-08-14 at 14 37 05

screen shot 2018-08-14 at 14 56 40

**Comment by jasmussen** [comment]
Normalized focus style:

screen shot 2018-08-14 at 15 06 44

**Comment by jasmussen** [comment]
In https://github.com/WordPress/gutenberg/pull/8916/commits/5874f5ba395df9eff02ebc55c7038506b14866d4 I pushed a little polish to focus and hover styles of blocks:

iterations

**Comment by mtias** [comment]
Oh, better structure for the scrollbar would also be needed.

**Comment by jasmussen** [comment]
I can fix it tomorrow so the scrollbar is right on the edge of the window as opposed to floating in the middle, if that's what you mean? I'll be sure to fix the padding also.

**Comment by jasmussen** [comment]
I pushed a change to the scrollable area:

screen shot 2018-08-16 at 08 29 55

screen shot 2018-08-16 at 08 30 24

Note I also changed the spacing from using 13px which didn't really fit with anything, to using 16px. I stored that as a new grid variable ($grid-size-large, with $grid-size being 8px), which we can use elsewhere to gain consistency.

**Comment by jasmussen** [comment]
screen shot 2018-08-16 at 11 13 59

Okay, I think I _actually_ fixed the scrollbar.

**Comment by jasmussen** [comment]
Fixed the switcher as well:

screen shot 2018-08-16 at 11 36 08

**Comment by jasmussen** [comment]
Fixed style variations and focus styles.

screen shot 2018-08-16 at 11 49 23

**Comment by paulwilde** [comment]
> Fixed the switcher as well:

The label text (Quote) looks too far to the bottom of the hover background in this image.

**Comment by chrisvanpatten** [comment]
I definitely like the direction this is going but all that padding on the outside edge just feels weird vs having it extend full-width. It creates a weirdly raggedy edge…

_untitled____add_new_page_ _mindful_ _wordpress

vs

_untitled____add_new_page_ _mindful_ _wordpress

But I do like the new icons (feels super weird for WordPress to use Material Icons instead of a WordPress-specific icon set, but they still work) and the lack of a background on the default icons isn't as bothersome in practice as I thought it might be. Overall looks good!

**Comment by youknowriad** [comment]
I pushed a refactoring to BlockIcons rendering which hopefully 🤷‍♂️  does:

 - Fix adding embed blocks (always use `BlockIcon` to render icons)
 - Centralize the CSS to apply colors in `BlockIcon` using the `showColors` prop
 - Consistent rendering for `BlockIcon` with or without colors (always have a `.editor-block-icon` wrapper)
 - Remove contrast checking for block icons (fix e2e tests)

**Comment by youknowriad** [comment]
So tests pass, I think we're ready to ship. Could anyone test everything for last minute breakage?

@chrisvanpatten I'm not designer but I had the same thoughts, I feel without the padding, it's better but that's shouldn't block us here. It's still a huge improvement over what we have, we can merge and iterate.

**Review by youknowriad — APPROVED** [review]
LGTM 👍

**Comment by chrisvanpatten** [comment]
Only remaining thing I noticed, but shouldn't be a blocker, is that dashicons now render blurry in non-retina displays (they may also be blurry in Retina, but I expect not because pixels!) because they're upscaled. It'd be nice if there was a fallback that just padded the dashicons instead of upscaling. But again, that shouldn't be a blocker.

File: test/e2e/specs/block-icons.test.js [hunk 2]
```diff
@@ -130,7 +130,7 @@ describe( 'Correctly Renders Block Icons on Inserter and Inspector', () => {
 		it( 'Renders the icon in the inserter with the correct background color and an automatically compute readable foreground color', async () => {
 			await searchForBlock( blockTitle );
 			validateSvgIcon( await getFirstInserterIcon() );
-			expect( await getBackgroundColor( INSERTER_ICON_SELECTOR ) ).toEqual( 'rgb(1, 0, 0)' );
+			expect( await getBackgroundColor( INSERTER_ICON_SELECTOR ) ).toEqual( 'rgba(0, 0, 0, 0)' );
```
Comment (jorgefilipecosta) [inline test/e2e/specs/block-icons.test.js]: The background color is in fact rgb(1, 0, 0) and with this change we don't verify if the background was in fact applied. I think we may need to update the selector const INSERTER_ICON_SELECTOR = `${ INSERTER_BUTTON_SELECTOR } .editor-block-icon`; to be  `${ INSERTER_BUTTON_SELECTOR } .editor-block-types-list__item-icon`.

Comment (youknowriad) [reply test/e2e/specs/block-icons.test.js]: Oh gotcha I'll take a look

Comment (jorgefilipecosta) [reply test/e2e/specs/block-icons.test.js]: .editor-block-icon is a child of .editor-block-types-list__item-icon and it seems colors are applied in .editor-block-types-list__item-icon.

Comment (youknowriad) [reply test/e2e/specs/block-icons.test.js]: It should be fixed.

**Review by karmatosed — APPROVED** [review]
I think we should ship what we have right now. The indent visually does help for padding on scrolling. Let's get in as we have now and we can always iterate.

**Comment by youknowriad** [comment]
Awesome work everyone, I love those PRs where so many people push code and review :P

File: packages/editor/src/components/block-icon/index.js [hunk 3]
```diff
@@ -25,18 +25,25 @@ function renderIcon( icon ) {
...
-				style={ {
-					backgroundColor: icon && icon.background,
-					color: icon && icon.foreground,
-				} }
-				className={ classnames( 'editor-block-icon__colors', className ) }
-			>
-				{ renderedIcon }
-			</div>
-		);
+	const style = showColors ? {
```
Comment (aduth) [inline packages/editor/src/components/block-icon/index.js]: Wasted effort to assign variable when following condition shortcuts early return without using (should be assigned after condition).

**Review by aduth — COMMENTED** [review]
Curious the impact this might have on bundle size. Probably not critical impact, but I'm sure it's a bump.

**Comment by afercia** [comment]
With this change, all the new hardcoded SVG icons miss the attributes required for accessibility:
`role="img" aria-hidden="true" focusable="false"` 😞Will open a new issue.

**Comment by jasmussen** [comment]
You can assign that issue to me.

**Pull request merged**